torch/_dynamo/graph_region_tracker.py

```diff
@@ -0,0 +1,290 @@
+        arg: Any
+        for arg in node.args:
+            if isinstance(arg, Node):
+                self._append((None, arg))
+
+        key: str
+        kwarg: Any
+        for key, kwarg in node.kwargs.items():
+            if isinstance(kwarg, Node):
+                self._append((key, kwarg))
+
+    def _append(self, name_arg: Tuple[Optional[str], Node]) -> None:
+        if self._cur == (None, None):
+            self._cur = name_arg
+        else:
+            self._queue.append(name_arg)
+
+
+class GraphRegionTracker:
+    def __init__(self) -> None:
+        self.loc_to_duplicates: Dict[str, IdenticalNodes] = defaultdict(list)
+        self.node_to_duplicates: Dict[Node, IdenticalNodes] = {}
+        self.node_to_global_state_hash: Dict[Node, int] = {}
+        self.input_pickler = InputPickler()
+
+    def _hash_node(
+        self, filename: str, lineno: int, instruction_pointer: Optional[int], node: Node
+    ) -> str:
+        from torch._inductor.codecache import sha256_hash
+
+        key = (
+            get_global_state_key(),
+            filename,
+            lineno,
+            instruction_pointer,
+            _normalize_args(node),
+        )
+        return sha256_hash(self.input_pickler.dumps(key))
+
+    def track_node(self, tx: "InstructionTranslatorBase", node: Node) -> None:
+        duplicates = self.loc_to_duplicates[
+            self._hash_node(
+                tx.f_code.co_filename, tx.lineno, tx.instruction_pointer, node
+            )
+        ]
+        duplicates.append(node)
+        self.node_to_duplicates[node] = duplicates
+
+    def is_identical(self, n0: Node, n1: Node) -> bool:
+        return (
+            n0 in self.node_to_duplicates
+            and n1 in self.node_to_duplicates
+            and self.node_to_duplicates[n0] == self.node_to_duplicates[n1]
+            and n0 is not n1
+        )
+
+    def get_identical_regions(self, graph: torch.fx.Graph) -> List[List[Region]]:
+        topological_ranking = {node: i for i, node in enumerate(graph.nodes)}
+        group_ranking = {}
+        region_groups = []
+
+        # Create region groups; a region group is a group
+        # of regions that are all identical. In this initial state
+        # each region in the group is a single node, and we discard
+        # groups that are only a single region.
+        # We track the topological ranking to start with groups later in the graph
+        # the reason for this is that we will necessarily create the largest groups first.
+        for group in self.loc_to_duplicates.values():
+            if len(group) > 1:
+                region_group = []
+                min_rank = math.inf
+                for node in group:
+                    min_rank = min(min_rank, topological_ranking[node])
+                    region_group.append([node])
+
+                region_groups.append(region_group)
+                group_ranking[id(region_group)] = min_rank
```

Comment:
nit: can we have region groups be a Tuple[region_group, min_rank] instead of using the id business? With id we have to worry about if the object is still alive

Reply:
Would this still allow constant time lookup? since we know the groups are alive for the duration of this function I'm not too worried.

Reply:
From discussion - store the ranking in the region in each region group itself